src/file_data.rs

```diff
@@ -0,0 +1,68 @@
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+use serde::Deserialize;
+
+use crate::data::{
+    DataSource, DataSourceInfo, EntryID, SlotMetaTile, SlotTile, SummaryTile, TileID,
+};
+use crate::http::schema::TileRequestRef;
+
+pub struct FileDataSource {
+    pub basedir: PathBuf,
+}
+
+impl FileDataSource {
+    pub fn new(basedir: impl AsRef<Path>) -> Self {
+        Self {
+            basedir: basedir.as_ref().to_owned(),
+        }
+    }
+
+    fn read_file<T>(&self, path: impl AsRef<Path>) -> T
+    where
+        T: for<'a> Deserialize<'a>,
+    {
+        let f = File::open(path).expect("opening file failed");
+        let f = zstd::Decoder::new(f).expect("zstd decompression failed");
+        let result = ciborium::from_reader(f).expect("cbor decoding failed");
+        result
```

Comment:
I'm still getting used to implicit returns (never been a huge fan tbh) but as long as they are the community standard, why not this:
```suggestion
        ciborium::from_reader(f).expect("cbor decoding failed");
```
Also my Rust naivete here but I don't see where `zstd` and `ciborium` are coming from

Reply:
Good catch.

Anything that is listed in the top-level `Cargo.toml` gets imported by its module name implicitly. Therefore, if you're happy using the syntax `module::function(...)` then you don't need an explicit `use` statement in the module.